src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs

```diff
@@ -120,18 +137,46 @@ public async Task StopAsync(CancellationToken cancelToken = default)
             }
         }
 
-        private int GetRandomPort()
+        private int GetLikelyAvailablePort()
         {
-            // Get a random port that has not already been attempted so we don't waste time trying
-            // to listen to a port we know is not free.
+            // Get an available port for use in the gRPC server. Try 4001 first, then select a random open port
+            // in the 30000-31000 range.
+            if (this.IsTcpPortFree(DefaultPort))
+            {
+                return DefaultPort;
+            }
+
+            int numAttempts = 50;
+
             int randomPort;
-            do
+            for (int i = 0; i < numAttempts; i++)
             {
                 randomPort = this.portGenerator.Next(MinPort, MaxPort);
+                if (this.IsTcpPortFree(randomPort))
+                {
+                    return randomPort;
+                }
             }
-            while (this.attemptedPorts.Contains(randomPort));
 
-            return randomPort;
+            throw new Exception(string.Format("Failed to get free port for local gRPC server after {0} attempts", numAttempts));
+        }
+
+        private bool IsTcpPortFree(int port)
+        {
+            var listener = new TcpListener(IPAddress.Loopback, port);
+            try
+            {
+                listener.Start();
+                return true;
+            }
+            catch (SocketException)
+            {
+                return false;
```

Comment:
Should we log a message if we fail to bind to a particular port? Previously we were doing this as part of the failure handling for the gRPC server, but it seems we're no longer likely to go down that path with this new implementation.

Reply:
Added a log for this with separate log text than the warning if we try to bind a gRPC server and get a conflict.